### PR TITLE
Migrated to use url-utils from Ghost-SDK

### LINKED ISSUE #10773: Extract URL Service utils into SDK
## Problem
URL Service utils are being used throughout the codebase in Ghost API/server and in the theme layer, which makes impossible to run the server without a theme layer.

## Solution
To decouple these utils from theme layer and make it a reusable utility library it should be extracted into Ghost-SDK in line how it's done in https://github.com/TryGhost/Ghost/issues/10618

LIst of changes:
- [x] extract `url-utils` package into [Ghost-SDK](https://github.com/TryGhost/Ghost-SDK) PR: https://github.com/TryGhost/Ghost-SDK/pull/114
- [x] decouple `config` and `settings cache` dependencies:
  - [x] config -  fields used:url , slugs , caching:301:maxAge , api:versions, admin:url
  - [x] settings cache - fields used: active_timezone
- [ ] integrate new package into Ghost (https://github.com/TryGhost/Ghost/pull/10787)

## Future improvements
As Ghost becomes more split into visible "frontend" and "server" parts these utilities might have a visible use case by only one or another, so these utils might benefit from splitting as well.

Nice-to-have:
- [x] make BASE_API_PATH and STATIC_IMAGE_URL_PREFIX configurable

refs https://github.com/TryGhost/Ghost/issues/10360

### PR DESCRIPTION
**Pull request by naz**
closes #10773

The main challenge during this extraction appeared unexpectedly - test suite ability to swap out `url` value in the Ghost instance config, which is not possible anymore with an external module.

After trying multiple approaches have settled on having a helper similar to `configUtils` (was used previously to swap out `url` in config). Took this approach over stubbing each method individually per test suite as it became apparent very quickly as it's just not worth the effort. For example, it's impossible to track which methods are used with which parameters in suites like `ghost_head_suite` (lots of helpers are being put together under test in one suite).  

@rishabhgrg @allouis  would love your opinion on this solution. There are loads of changes but mainly they are doing the same steps for each module: 
1. Change using `urlService.utils` to use `/lib/url-utils`
2. Correct any failing tests (usually had to apply the stubbing method described above)

Left to do:
- [x] fix regression tests

### CODE REVIEW

File: core/test/utils/urlUtils.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+const _ = require('lodash');
```
Comment (gargol) [inline core/test/utils/urlUtils.js]: This module is the core of how stubbing works when `url` has to be substituted.

**Comment by rishabhgrg** [comment]
@gargol LGTM 🎉 I just pushed a fix for linting 😄

**Review by rshbhgrg — APPROVED** [review]

File: core/test/utils/urlUtils.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    const stubInstance = getInstance(options);
+
+    Object.keys(urlUtils).forEach((key) => {
+        sandbox.stub(urlUtils, key).callsFake(stubInstance[key]);
+    });
+};
+
+// Method for regressions tests must be used with restore method
+const stubUrlUtilsFromConfig = () => {
+    sandbox = sinon.createSandbox();
```
Comment (allouis) [inline core/test/utils/urlUtils.js]: Is there a reason we create a new sandbox every time this is called? I think you can create as `const sandbox` and then replace this line w/ `sandbox.restore()`.

Comment (gargol) [reply core/test/utils/urlUtils.js]: @allouis wasn't sure about behavior in this case, but looks like you are right, works perfectly with a single instance :+1: This simplified a code a bit, niceeee :smiley:

**Comment by allouis** [comment]
@gargol I think this approach is good for now, though I do wonder if it is the kind of pattern we want to continue with.

I think the previous approach of the `configUtils` and setting config in tests was a nice pattern, as it is generic, and could be used all over the tests, rather than making a new "stubber module" for each piece of Ghost we pull into an SDK.

I'm thinking maybe we could update `configUtils` to clear the require cache of the modules that depend on `config` which would alleviate a lot of this stubbing, though it introduces some implicit behaviour, we could have it as opt-in, making it more explicit.

We could probably optimise the shit out of it, and keep track of dirty modules, but something like below _could_ work!

Say we had some module purger like this (i quickly tested this, it's probs buggy):
```js
function purge_module(id) {
  const purge_id = require.resolve(id);

  Object.keys(require.cache).forEach(function (cached_module_id) {
    const cached_module = require.cache[cached_module_id];

    if (cached_module && cached_module.children.map(mod => mod.id).includes(purge_id)) {
      purge_module(cached_module_id);
    }
  });
  require.cache[purge_id] = undefined;
}
```

If the `configUtils#set` method looked something like this:

```js
configUtils.set = function set() {
  purge_module('../path/to/config');
  const config = require('../path/to/config');

  // do setting of config
}
```

And the tests were written something like this:

```js
it('does stuff', function () {
  configUtils.set('whatever', 'stuff');
  
  const moduleToBeTested = require('./some/module');

  assert.equals(moduleToBeTested.method(), 'stuff');
});
```

Maybe this is mental - I'm not quite sure :shrimp:

This isn't something we need to do now, or ever - but I think a pattern should be worked out that we're happy with repeating a lot, if that makes sense? :relaxed:

**Review by allouis — APPROVED** [review]

File: core/test/utils/urlUtils.js (hunk 2 above)
Comment (allouis) [reply core/test/utils/urlUtils.js]: Yay! :dancer:

**Comment by gargol** [comment]
@allouis like how generic this approach with require.cache purging is and think that could be a path to explore once we have another module that depends on config being extracted :+1: Not sure how we'll manage all the requires in the tests though, as loads of modules are dependent on config and we'd have to re-require the purged ones over again in test cases that use the purging method? Or they'll resolve themselves automatically?

As a side note. Think that modifying config withing unit tests directly, as it's done now, spotted some flaw in how our tests are structured. Because when we are settings for example `config.set('url')` we suddenly modify the behavior in modules that are far in the dependency tree from the tested module (e.g. don't think it's a good practice mocking behavior of a child of a child module). For example the `ghost_head_spec` [suite](https://github.com/TryGhost/Ghost/blob/aaf856c644a8f2d7cbaffd5044b4498dd8220890/core/test/unit/helpers/ghost_head_spec.js). It is fine to modify 'referrerPolicy' or `useStructuredData` values, but modifying `url` was making a change to something that module under test didn't even have a notion of :confused: Even now, using `urlUtils` there felt a bit wrong, but rewriting whole test suite is not in scope of this PR. Think if we spot places where we have to create stubs like url-utils again, we should think again if tests are not knowing too much. This would avoid creating new *-utils alltogether and try mocking direct dependencies of that module instead :thinking:

**Pull request merged**